### PR TITLE
Make priority granularity configurable #599

### LINKED ISSUE #599: Make priority granularity configurable
I think it would be good if the priority granularity could be configurable. ecChronos calculates priority based on amount of hours that have passed since the repair "could" run. What this effectively means is that all schedules executed the same hour will have the same priority, i.e having repair interval less than 1 hour will produce strange results.

What i propose is to expose the granularity as a global setting in ecc.yml, something like:

```
repair:
  priority:
    granularity_unit: HOURS
```

The possible values for granularity_unit should be HOURS, MINUTES or SECONDS. HOURS should be default to not break upgrade from older versions of ecChronos. Also it should be mentioned as a comment in the config file that repair must be paused when changing the granularity. Additionally the priority calculation should be looked over to make sure we cap priority (overflow protection).

### PR DESCRIPTION
**Pull request by SajidRiaz138**
Closes #599

### CODE REVIEW

File: application/src/main/resources/ecc.yml [hunk 1]
```diff
@@ -86,6 +86,15 @@ repair:
     time: 7
     unit: days
   ##
+  ##  The unit of time granularity for priority calculation, can be HOURS, MINUTES, or SECONDS.
+  ##  It determines how often the priority is recalculated.
```
Comment (masokol) [inline application/src/main/resources/ecc.yml:90]: "It determines how often the priority is recalculated." this is not true. It's used to calculate priority not "how often".

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduledJob.java [hunk 2]
```diff
@@ -172,9 +173,16 @@ public final int getRealPriority(final long lastSuccessfulRun)
             return -1;
         }
 
-        int hours = (int) (diff / ONE_HOUR_IN_MILLIS) + 1;
+        long granularityInMs = TimeUnit.MILLISECONDS.convert(1, myGranularityUnit);
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduledJob.java:176]: myGranularityUnit.toMillis(1);

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TestTableRepairJob.java [hunk 3]
```diff
@@ -664,6 +664,52 @@ public void testRunnableWithOffset() {
         assertThat(myRepairJob.runnable()).isFalse();
     }
 
+    @Test
+    public void testGetRealPriorityWithMinuteGranularity()
+    {
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TestTableRepairJob.java:669]: Add a testcase for overflow.

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduledJob.java [hunk 4]
```diff
@@ -313,11 +322,17 @@ public static class Configuration
          */
         public final long backoffInMs;
 
+        /**
+         * The unit of time granularity used for priority calculation in scheduling jobs.
+         */
+        public final TimeUnit granularityUnit;
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduledJob.java:328]: priorityGranularity

File: application/src/main/resources/ecc.yml [hunk 5]
```diff
@@ -86,6 +86,15 @@ repair:
     time: 7
     unit: days
   ##
+  ##  The unit of time granularity for priority calculation, can be HOURS, MINUTES, or SECONDS.
+  ##  This unit is used in the calculation of priority.
+  ##  Default is HOURS for backward compatibility.
+  ##  Ensure to pause repair operations when changing the granularity to maintain consistency.
```
Comment (jwaeab) [inline application/src/main/resources/ecc.yml:92]: 'when' should probably be something as 'prior to', just to be crystal clear of when to do it.

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairConfiguration.java [hunk 6]
```diff
@@ -14,13 +14,15 @@
...
+import static com.ericsson.bss.cassandra.ecchronos.core.repair.RepairOptions.RepairParallelism;
+
 import java.util.Objects;
 import java.util.concurrent.TimeUnit;
 
 /**
  * Configuration options for table repairs.
  */
-@SuppressWarnings("FinalClass")
+@SuppressWarnings ("FinalClass")
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairConfiguration.java:25]: Why a space in between?

File: docs/UPGRADE.md [hunk 7]
```diff
@@ -33,6 +33,14 @@ directly within the `ecc.yaml` file.
 
 It’s important to note that while making these changes, the default behavior remains unchanged.
 
+## Priority Calculation
+
+The unit of time granularity for priority calculation, can be HOURS, MINUTES, or SECONDS.
+This unit is used in the calculation of priority. Default is HOURS for backward compatibility.
+Ensure to pause repair operations when changing the granularity to maintain consistency.
```
Comment (jwaeab) [inline docs/UPGRADE.md:40]: Same comment as for 'ecc.yml'.

File: docs/UPGRADE.md [hunk 8]
```diff
@@ -33,6 +33,14 @@ directly within the `ecc.yaml` file.
 
 It’s important to note that while making these changes, the default behavior remains unchanged.
 
+## Priority Calculation
+
+The unit of time granularity for priority calculation, can be HOURS, MINUTES, or SECONDS.
+This unit is used in the calculation of priority. Default is HOURS for backward compatibility.
+Ensure to pause repair operations when changing the granularity to maintain consistency. 
+HOURS  # or MINUTES, SECONDS. It’s important to note that while making these changes,
```
Comment (jwaeab) [inline docs/UPGRADE.md:41]: This line looks strange. Copy+paste problem?

**Review by jwaeab — CHANGES_REQUESTED** [review]

**Review by jwaeab — APPROVED** [review]

**Review by jwaeab — APPROVED** [review]

**Pull request merged**